### PR TITLE
[wasm][bcl][websockets] Add WebSocket support.

### PR DESCRIPTION
**Pull request by kjpou1**
Add the ability to initialize a `WebAssembly.Net.WebSockets::ClientWebSocket` from `System.Net.WebSockets.ClientWebSocket`.

`System.Net.WebSockets.ClientWebSocket` will delegate all calls to the WebAssembly ClientWebSocket.  

- Add support for marshaling `Uri` class correctly.  This was a cause of failures.
- Add bindings tests for `Uri` marhaling.

### CODE REVIEW

**Comment by kjpou1** [comment]
@monojenkins build failed

File: mcs/class/System/System.Net.WebSockets/WebSocketHandle.wasm.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        public Task CloseAsync(WebSocketCloseStatus closeStatus, string statusDescription, CancellationToken cancellationToken) =>
+            _webSocket.CloseAsync(closeStatus, statusDescription, cancellationToken);
+
+        public Task CloseOutputAsync(WebSocketCloseStatus closeStatus, string statusDescription, CancellationToken cancellationToken) =>
+            _webSocket.CloseOutputAsync(closeStatus, statusDescription, cancellationToken);
+
+        public async Task ConnectAsyncCore(Uri uri, CancellationToken cancellationToken, ClientWebSocketOptions options)
+        {
+            // TODO #14480 : Not currently implemented, or explicitly ignored:
+            // - ClientWebSocketOptions.UseDefaultCredentials
```
Comment (kg) [inline mcs/class/System/System.Net.WebSockets/WebSocketHandle.wasm.cs]: We should detect the user providing non-default values for any of these unimplemented options and throw

Comment (kg) [reply mcs/class/System/System.Net.WebSockets/WebSocketHandle.wasm.cs]: In particular, requesting a proxy and silently performing an unproxied connection can have significant negative consequences

File: sdks/wasm/framework/src/WebAssembly.Bindings/LinkDescriptor/WebAssembly.Bindings.xml [hunk 2]
```diff
@@ -28,6 +28,7 @@
             <method name="ObjectToString" />
             <method name="GetDateValue" />
             <method name="CreateDateTime" />
+            <method name="CreateUri" />
```
Comment (kg) [inline sdks/wasm/framework/src/WebAssembly.Bindings/LinkDescriptor/WebAssembly.Bindings.xml]: Add GetUriType to this list also since you added it along with CreateUri

File: sdks/wasm/framework/src/WebAssembly.Framework.sln [hunk 3]
```diff
@@ -7,7 +7,7 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WebAssembly.Bindings", "Web
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WebAssembly.Net.Http", "WebAssembly.Net.Http\WebAssembly.Net.Http.csproj", "{D721826B-CD18-4FBC-A3BC-7CDDEA41F3A2}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WebAssembly.Net.WebSockets", "WebAssembly.Net.WebSockets\WebAssembly.Net.WebSockets.csproj", "{D399ED16-6F8E-4D6D-BF44-434D5F2986D7}"
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WebAssembly.Net.WebSockets", "WebAssembly.Net.WebSockets\WebAssembly.Net.WebSockets.csproj", "{5080984C-B88B-4DB7-A2A9-9A6C6DA5382C}"
```
Comment (kg) [inline sdks/wasm/framework/src/WebAssembly.Framework.sln]: Why did this GUID change?

File: sdks/wasm/framework/src/WebAssembly.Net.WebSockets/ClientWebSocket.cs [hunk 4]
```diff
@@ -540,106 +531,107 @@ public async Task<T> DequeuePayloadAsync (CancellationToken cancellationToken =
...
 		}
 
-		public sealed class ClientWebSocketOptions {
+		public sealed class ClientWebSocketOptions
+		{
 			private bool isReadOnly; // After ConnectAsync is called the options cannot be modified.
 			private readonly IList<string> requestedSubProtocols;
 
-			internal ClientWebSocketOptions ()
+			internal ClientWebSocketOptions()
```
Comment (kg) [inline sdks/wasm/framework/src/WebAssembly.Net.WebSockets/ClientWebSocket.cs]: A bunch of formatting damage happened in this file, it should be reverted and if the file needs a formatting cleanup let's do that in a separate PR so the git blame is less confusing

File: sdks/wasm/src/driver.c [hunk 5]
```diff
@@ -488,6 +489,28 @@ class_is_task (MonoClass *klass)
 	return 0;
 }
 
+MonoClass* mono_get_uri_class(MonoException** exc) 
+{
+	MonoAssembly* assembly = mono_wasm_assembly_load ("WebAssembly.Bindings");
+	if (!assembly)
+		return NULL;
+	MonoClass* klass = mono_wasm_assembly_find_class(assembly, "WebAssembly", "Runtime");
```
Comment (kg) [inline sdks/wasm/src/driver.c]: If we're already going to the trouble here of looking up a class in an assembly and calling a method, why don't we just look System.Uri up directly?

File: mcs/class/System/System.Net.WebSockets/WebSocketHandle.wasm.cs (hunk 1 above)
Comment (kg) [reply mcs/class/System/System.Net.WebSockets/WebSocketHandle.wasm.cs]: You can consider this suggestion optional for now, we just need to do it eventually

File: sdks/wasm/framework/src/WebAssembly.Bindings/LinkDescriptor/WebAssembly.Bindings.xml (hunk 2 above)
Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Bindings/LinkDescriptor/WebAssembly.Bindings.xml]: `GetUriType` is marked as public so should not need it.

File: sdks/wasm/src/driver.c (hunk 5 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: I tried looking it up with `mono_class_from_name (mono_get_corlib(), "System", "Uri");` but no class is found.

File: mcs/class/System/System.Net.WebSockets/WebSocketHandle.wasm.cs (hunk 1 above)
Comment (kjpou1) [reply mcs/class/System/System.Net.WebSockets/WebSocketHandle.wasm.cs]: The `ClientWebSocketOptions` throws not implemented for these.

File: sdks/wasm/framework/src/WebAssembly.Framework.sln (hunk 3 above)
Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Framework.sln]: Yes, because I removed it and added it back.  Will change this.

**Comment by kjpou1** [comment]
@monojenkins build

File: sdks/wasm/src/driver.c (hunk 5 above)
Comment (kg) [reply sdks/wasm/src/driver.c]: That might be because it doesn't always live in mscorlib.dll, it looks like according to the docs site it can live in System or System.Runtime instead.

**Comment by kjpou1** [comment]
@monojenkins build failed

File: sdks/wasm/framework/src/WebAssembly.Net.WebSockets/ClientWebSocket.cs (hunk 4 above)
Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Net.WebSockets/ClientWebSocket.cs]: Addressed in recent commits

File: sdks/wasm/framework/src/WebAssembly.Framework.sln (hunk 3 above)
Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Framework.sln]: Addressed in recent commits

File: sdks/wasm/src/driver.c (hunk 5 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: Addressed in recent commits

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by akoeplinger** [comment]
@monojenkins build Linux x64 All Profiles Build

**Comment by kjpou1** [comment]
@monojenkins build Project Files Diff

**Comment by kjpou1** [comment]
@monojenkins build Linux x64 All Profiles Build

**Comment by akoeplinger** [comment]
@monojenkins commit csproj
